README.md
# Oraq

## Description

Oraq (Ordered redis asynchronous queue) is a cross process, cross server, fast, reliable and lightweight redis-based async queue for Node.

## Features

- super fast
- cross process
- easy to use

## Compatibility

Supports node.js >= 7.6.0.

## Installation

```bash
npm install --save oraq
```

## Usage

```js
const Oraq = require('oraq');
const oraq = new Oraq({
  id: 'myQueue',
  concurrency: 1
});

const job = ms => new Promise(resolve => setTimeout(() => resolve(ms), ms));

// do a lot of heavy jobs with concurrency 1
Promise.all([
  14000,
  8000,
  10000,
  7000
].map(delay => oraq.limit(job, {jobData: delay})))
  .then(jobResults => console.log(jobResults));
```

## Constructor options

```js
const oraq = new Oraq({
  id,                        // id {string} (limiters with the same prefix and id share their queues, "queue" by default)
  prefix,                    // custom redis key prefix {string} ("oraq" by default)
  connection,                // redis connection param {any} (anything that ioredis constructor supports)
  ping,                      // processing job keep alive interval in ms {integer} (60000 by default)
  timeout,                   // job will run after this time {integer} (in case of too long previous tasks processing, 2 * 60 * 60 * 1000 (2 hours) by default)
  concurrency,               // jobs concurrency {integer} (1 by default)
  mode                       // mode {string} ("limiter" - rate limiter (no order guarantee) or "queue" - real queue (keep order), "queue" by default)
});
```

## API

```js
const oraq = new Oraq();

oraq.limit(
  job,                       // job {function}
  jobOptions = {             // job options {object}
    jobId,                   // job id {string} (random string by default)
    jobData,                 // data to pass as an argument to the job {any}
    lifo                     // last-in-first-out {boolean} (false by default)
  }
).then(jobResult => console.log(jobResult));

oraq.quit()                  // same as ioredis quit method (used to close redis connection)
  .then(() => console.log('Redis connection is closed'));

oraq.removeJobById('jobId')  // remove job from pending queue
  .then(() => console.log('Job is removed'));
```

## Tests

```bash
npm test
```

## License

MIT © [Yevhen Samoilenko](https://github.com/deugene)


CHANGELOG.md
# Changelog

All notable changes are documented in this file.

## 0.2.4

- keyevent listener fix

## 0.2.3

### Added

- mode option

## 0.2.2

### Added

- coordinator options validation

### Changed

- lock keys format
- improve coordinator logic

## 0.2.0

### Added

- removeJobById method

### Changed

- update readme

## 0.1.0

- initial release


test/index.js
'use strict';

const assert = require('assert');
const Oraq = require('../index');
const Coordinator = require('../Coordinator');


/**
 *
 * Coordinator tests
 *
 */

/**
 * options validation
 */
(() => {
  assert.throws(() => new Coordinator(), /^Error: options are required/);
  assert.throws(() => new Coordinator({}), /^Error: jobId/);
  assert.throws(() => new Coordinator({
    jobId: '1'
  }), /^Error: keyPending/);
  assert.throws(() => new Coordinator({
    jobId: '1',
    keyPending: 'pending'
  }), /^Error: keyProcessing/);
  assert.throws(() => new Coordinator({
    jobId: '1',
    keyPending: 'pending',
    keyProcessing: 'processing'
  }), /^Error: lock/);
  assert.throws(() => new Coordinator({
    jobId: '1',
    keyPending: 'pending',
    keyProcessing: 'processing',
    lock: ':lock'
  }), /^Error: concurrency/);
  assert.throws(() => new Coordinator({
    jobId: '1',
    keyPending: 'pending',
    keyProcessing: 'processing',
    lock: ':lock',
    concurrency: 1
  }), /^Error: timeout/);
  assert.throws(() => new Coordinator({
    jobId: '1',
    keyPending: 'pending',
    keyProcessing: 'processing',
    lock: ':lock',
    concurrency: 1,
    timeout: 10000
  }), /^Error: client/);
})();

/**
 * properties
 */

(() => {
  const coordinator = new Coordinator({
    jobId: '1',
    keyPending: 'pending',
    keyProcessing: 'processing',
    lock: ':lock',
    concurrency: 1,
    timeout: 10000,
    client: {}
  });

  assert.ok(coordinator);
  assert.ok(coordinator.canRun, true);
  assert.strictEqual(typeof coordinator.keepAlive, 'function');
  assert.strictEqual(typeof coordinator.stopKeepAlive, 'function');
  assert.strictEqual(typeof coordinator.wait, 'function');
  assert.strictEqual(typeof coordinator.stopWait, 'function');
})();


/**
 *
 * Oraq tests
 *
 */

Promise.resolve().then(async () => {
  /**
   * properties
   */

  const oraq = new Oraq();

  try {
    assert.ok(oraq);
    assert.strictEqual(typeof oraq.limit, 'function');
    assert.strictEqual(typeof oraq.quit, 'function');
    assert.strictEqual(typeof oraq.removeJobById, 'function');
  } finally {
    await oraq.quit();
  }
}).then(async () => {
  /**
   * processing time
   */

  const job = ms => new Promise(resolve => setTimeout(() => resolve(ms), ms));
  const oraq = new Oraq({
    id: 'test',
    prefix: 'oraq_test',
    concurrency: 1
  });

  try {
    const start = Date.now();

    await Promise.all([
      1000,
      1000,
      1000,
      1000
    ].map(delay => oraq.limit(job, {jobData: delay})));

    const end = Date.now();

    assert.strictEqual(end >= start + 4000, true);
  } finally {
    // close redis connections
    await oraq.quit();
  }
});


Coordinator.js
'use strict';

/**
 * Class representing a coordinator for Oraq
 */
class Coordinator {
  /**
   * Creates an instance of Coordinator
   *
   * @param   {object}   options
   * @param   {string}   options.jobId
   * @param   {object}   options.client          redis client
   * @param   {number}   options.concurrency     jobs concurrency
   * @param   {string}   options.keyPending      pending queue key
   * @param   {string}   options.keyProcessing   processing queue key
   * @param   {number}   options.timeout         job will run after this time (in case of too long previous tasks processing)
   * @param   {string}   options.mode            mode {string} ('limiter' - rate limiter (no order guarantee); 'queue' - real queue (keep order))
   * @memberof Coordinator
   */
  constructor(options) {
    this._validate(options);

    const {jobId, client, concurrency, keyPending, keyProcessing, timeout, lock, mode} = options;

    this._jobId = jobId;
    this._keepAliveTimeout = null;
    this._waitTimeout = null;
    this._concurrency = concurrency;
    this._client = client;
    this._resolve = null;
    this._keyPending = keyPending;
    this._keyProcessing = keyProcessing;
    this._timeout = timeout;
    this._lock = lock;
    this._mode = mode;
    this._startTime = null;
    this._canRun = new Promise(resolve => this._resolve = resolve);
  }

  /**
   * Validate options
   *
   * @param   {object}   options
   * @private
   * @memberof Coordinator
   */
  _validate(options) {
    if (!options) {
      throw new Error('options are required');
    }
    if (typeof options.jobId !== 'string') {
      throw new Error('jobId must be a string');
    }
    if (typeof options.keyPending !== 'string') {
      throw new Error('keyPending must be a string');
    }
    if (typeof options.keyProcessing !== 'string') {
      throw new Error('keyProcessing must be a string');
    }
    if (typeof options.lock !== 'string') {
      throw new Error('lock must be a string');
    }
    if (!Number.isInteger(options.concurrency)) {
      throw new Error('concurrency must be an integer');
    }
    if (!Number.isInteger(options.timeout)) {
      throw new Error('timeout must be an integer');
    }
    if (!options.client) {
      throw new Error('client is required');
    }
  }

  get canRun() {
    return this._canRun;
  }

  /**
   * Keep lock key alive
   *
   * @param   {number}   ms
   * @memberof Coordinator
   */
  async keepAlive(ms) {
    const lockKey = `${this._keyProcessing}:${this._jobId}${this._lock}`;

    this.stopKeepAlive();
    await this._client.setex(lockKey, ms * 2 / 1000, '').catch(() => null);
    this._keepAliveTimeout = setTimeout(() => this.keepAlive(ms), ms);
  }

  /**
   * Clear keep alive timeout
   *
   * @memberof Coordinator
   */
  stopKeepAlive() {
    if (this._keepAliveTimeout) {
      clearTimeout(this._keepAliveTimeout);
      this._keepAliveTimeout = null;
    }
  }

  /**
   * Wait for queue changes
   *
   * @param   {number}   ms
   * @memberof Coordinator
   */
  async wait(ms) {
    this.stopWait();
    await this._setCanRun();
    this._waitTimeout = setTimeout(() => this.wait(ms), ms);
  }

  /**
   * Clear wait timeout
   *
   * @memberof Coordinator
   */
  stopWait() {
    if (this._waitTimeout) {
      clearTimeout(this._waitTimeout);
      this._waitTimeout = null;
    }
  }

  /**
   * Check if job can run and fulfill canRun promise
   *
   * @private
   * @memberof Coordinator
   */
  async _setCanRun() {
    if (!this._startTime) {
      this._startTime = Date.now();
    }
    if (Date.now() - this._startTime > this._timeout) {
      this._resolve();
      return;
    }

    await this._removeStuckJobs(this._keyPending);
    await this._removeStuckJobs(this._keyProcessing);

    const [llenRes, lindexRes] = await this._client
      .multi()
      .llen(this._keyProcessing)
      .lindex(this._keyPending, -1)
      .exec();
    const processingCount = llenRes[1];
    const nextJobId = lindexRes[1];

    if (processingCount < this._concurrency) {
      switch (this._mode) {
        case 'limiter':
          this._resolve();
          break;
        case 'queue':
          nextJobId === this._jobId && this._resolve();
          break;
      }
    }
  }

  /**
   * Remove stuck jobs
   *
   * @param   {string}   queueKey
   * @returns {Promise}
   * @private
   * @memberof Coordinator
   */
  async _removeStuckJobs(queueKey) {
    const jobIds = await this._client.lrange(queueKey, 0, -1);
    let stuckJobs = new Set();

    if (!jobIds.length) {
      return;
    }

    for (const jobId of jobIds) {
      if (jobId) {
        const existing = await this._client.exists(`${queueKey}:${jobId}${this._lock}`);

        if (!existing) {
          stuckJobs.add(jobId);
        }
      }
    }

    if (!stuckJobs.size) {
      return;
    }

    const pipe = this._client.multi();

    for (const jobId of stuckJobs) {
      pipe.lrem(queueKey, 0, jobId);
    }

    return pipe.exec().catch(() => null);
  }
}

module.exports = Coordinator;


index.js
'use strict';

const Redis = require('ioredis');
const crypto = require('crypto');
const {promisify} = require('util');
const randomBytes = promisify(crypto.randomBytes);
const {name: packageName} = require('./package.json');
const Coordinator = require('./Coordinator');

const DEFAULT_ID = 'queue';

/**
 * Class representing an ordered redis asynchronous queue
 */
class Oraq {
  /**
   * Creates an instance of Oraq
   *
   * @param   {object}   options
   * @param   {string}   options.prefix        custom redis key prefix
   * @param   {string}   options.id            id (limiters with the same prefix and id share their queues)
   * @param   {*}        options.connection    redis connection param
   * @param   {number}   options.concurrency   jobs concurrency
   * @param   {number}   options.ping          processing job keep alive interval
   * @param   {number}   options.timeout       job will run after this time (in case of too long previous tasks processing)
   * @param   {string}   options.mode          mode {string} ('limiter' - rate limiter (no order guarantee); 'queue' - real queue (keep order))
   * @memberof Oraq
   */
  constructor(options) {
    const {
      id = DEFAULT_ID,
      prefix = packageName,
      connection,
      ping = 60 * 1000,              // 1 minute
      timeout = 2 * 60 * 60 * 1000,  // 2 hours
      concurrency = 1,
      mode = 'queue'
    } = options || {};

    this._mode = mode;
    this._ping = ping;
    this._timeout = timeout;
    this._concurrency = concurrency;
    // keys for data store
    this._key = [prefix, id].join(':');
    this._keyProcessing = this._key + ':processing';
    this._keyPending = this._key + ':pending';
    // lock key suffix
    this._lock = ':lock';
    // redis client
    this._client = new Redis(connection);
    // enable namespace events
    this._client.config('SET', 'notify-keyspace-events', 'Kgxl');
    // keyspace events subscriber
    this._subscriber = new Redis(connection);
    this._ready = null;
  }

  /**
   * Initialize
   *
   * @private
   * @memberof Oraq
   */
  async _init() {
    if (this._ready === null) {
      this._ready = new Promise((resolve, reject) => {
        this._subscriber.psubscribe(`__keyspace@0__:${this._key}*`, err => {
          if (err) {
            reject(err);
          } else {
            resolve();
          }
        });
      });
    }

    return this._ready;
  }

  /**
   * Limit jobs rate
   *
   * @param   {function}  job       job to process
   * @param   {object}    options
   * @param   {string}    options.jobId
   * @param   {*}         options.jobData   data to pass to the job
   * @param   {boolean}   options.lifo      last-in-first-out (fifo by default)
   * @memberof Oraq
   */
  async limit(job, {jobId, jobData, lifo = false} = {}) {
    if (!this._isSubscribed) {
      await this._init();
    }

    jobId = jobId || (await randomBytes(16)).toString('hex');

    const coordinator = new Coordinator({
      jobId,
      client: this._client,
      concurrency: this._concurrency,
      keyPending: this._keyPending,
      keyProcessing: this._keyProcessing,
      timeout: this._timeout,
      lock: this._lock,
      mode: this._mode
    });
    const onKeyEvent = this._getOnKeyEvent(coordinator);

    try {
      let result;

      // add job to the pending queue
      await this._client
        .multi()
        .setex(`${this._keyPending}:${jobId}${this._lock}`, this._timeout * 1.5 / 1000, '')
        [lifo ? 'rpush' : 'lpush'](this._keyPending, jobId)  // eslint-disable-line no-unexpected-multiline
        .exec();
      // listen processing key events
      this._subscriber.addListener('pmessage', onKeyEvent);
      // concurrency
      await coordinator.wait(this._ping);
      await coordinator.canRun;
      this._subscriber.removeListener('pmessage', onKeyEvent);
      coordinator.stopWait();
      // create lock key and keep it alive
      await coordinator.keepAlive(this._ping);
      // move job from pending to processing queue
      await this._client.multi()
        .lrem(this._keyPending, 1, jobId)
        .lpush(this._keyProcessing, jobId)
        .del(`${this._keyPending}:${jobId}${this._lock}`)
        .exec();
      // run job
      result = await job(jobData);

      return result;
    } finally {
      // stop all timers and remove all listeners
      coordinator.stopKeepAlive();
      coordinator.stopWait();
      this._subscriber.removeListener('pmessage', onKeyEvent);
      // remove processing job id and lock key
      await this._client
        .multi()
        .lrem(this._keyProcessing, 1, jobId)
        .del(`${this._keyProcessing}:${jobId}${this._lock}`)
        .exec();
    }
  }

  /**
   * Get on key event listener
   *
   * @param   {object}   coordinator
   * @returns {function}
   * @private
   * @memberof Oraq
   */
  _getOnKeyEvent(coordinator) {
    return (pattern, channel, message) => {
      if (message === 'expired' && channel.endsWith(this._lock)) {
        for (const queueKey of [this._keyPending, this._keyProcessing]) {
          const queueStart = `__keyspace@0__:${queueKey}:`;

          if (channel.startsWith(queueStart)) {
            const expiredJobId = channel.slice(queueStart.length, -this._lock.length);

            this._client.lrem(queueKey, 1, expiredJobId).catch(console.error);
          }
        }
      } else if (message === 'lrem') {
        coordinator.wait(this._ping).catch(console.error);
      }
    };
  }

  /**
   * Close redis connections
   *
   * @returns {Promise}
   * @memberof Oraq
   */
  async quit() {
    await this._subscriber.quit();
    return this._client.quit();
  }

  /**
   * Close redis connections
   *
   * @param   {string}   jobId
   * @returns {Promise}
   * @memberof Oraq
   */
  async removeJobById(jobId) {
    return this._client
      .multi()
      .del(`${this._keyPending}:${jobId}${this._lock}`)
      .lrem(this._keyPending, 1, jobId)
      .exec();
  }
}


module.exports = Oraq;
module.exports.default = Oraq;
